File: Week_02/G20200343030632/Leetcode_226_632.java
import java.util.Deque;
import java.util.LinkedList;


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode invertTree(TreeNode root) {
        // 迭代
        Deque<TreeNode> stack = new LinkedList<>();
        stack.addFirst(root);
        while (!stack.isEmpty()) {
            TreeNode cur = stack.removeFirst();
            if (cur != null) {
                TreeNode temp = cur.left;
                cur.left = cur.right;
                cur.right = temp;
                stack.addFirst(cur.left);
                stack.addFirst(cur.right);

            }
        }
        return root;
    }
}

File: Week_02/G20200343030632/Leetcode_226_632.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def invertTree(self, root: TreeNode) -> TreeNode:
        # return self.invertTree_v1(root)
        # return self.invertTree_v2(root)
        return self.invertTree_v3(root)
    
    def invertTree_v1(self, root: TreeNode) -> TreeNode:
        self.helper(root)
        return root
    
    def helper(self, node: TreeNode):
        if node:
            node.left, node.right = node.right, node.left
            self.helper(node.left)
            self.helper(node.right)

    def invertTree_v2(self, root: TreeNode) -> TreeNode:
        if root:
            invert = self.invertTree_v2
            root.left = invert(root.right)
            root.right = invert(root.left)
            return root
    
    def invertTree_v3(self, root: TreeNode) -> TreeNode:
        # 迭代
        stack = [root]
        while stack:
            cur = stack.pop()
            if cur:
                cur.left, cur.right = cur.right, cur.left
                stack.append(cur.left)
                stack.append(cur.right)
        return root